Enable downcasting of inner err field to Printable

omkar-mohanty opened the pull request:
Fixes #39 

I initially attempted to upcast using `&&(*self.err) as &dyn Any`, but the borrow checker raised lifetime issues with the reference.
To work around this, I introduced a wrapper trait combining `std::error::Error` and `DowncastSync` (`PlironError`). This allows us to downcast the inner error via `as_any()` and invoke `Printable::disp` when applicable.

Let me know if this approach looks good or if you'd prefer a different direction.

vaivaswatha commented:
I'm not sure I like the idea of using a wrapper trait, but if there's no other way, it may be okay.

But more importantly, `if let Some(self_val) = (*self.err).as_any().downcast_ref::<&dyn Printable>()` this will not work I think (maybe we can verify with a test). You'll need to use the `trait_cast` utility I wrote, to register the concrete type that implements `Printable` in that utility.

omkar-mohanty commented:
Got it. I did end up using `trait_cast` as suggested, but I had to keep the wrapper trait (`PlironError`) because the borrow checker wasn’t happy with the direct upcast attempts — it kept complaining about lifetimes when trying things like `&&(*self.err) as &dyn Any`.

Let me know if this approach seems okay or if you'd prefer something else.

vaivaswatha commented:
It is ok I think. 

1. I think we should call the trait `AnyError` (ie some hint that it's a trait and not  type. There are already do many error types making it all confusing.
2. Can you add a test in the same file below, with aj error struct (or enum) that implements `Printable`, and is then wrapped by another `Error`, and ensure that the printing prints the inner one also correctly. Remember to also use the macro in `trait_cast` to mark that the type should be convertible to `Printable`.

omkar-mohanty commented:
I've renamed the trait to `AnyError`, added the test with a nested error that implements `Printable`, and verified that the inner error is printed correctly. Also I've added some docs for `AnyError`. Let me know if anything else needs tweaking.

vaivaswatha commented on src/result.rs:
```diff
@@ -63,7 +79,13 @@ impl Printable for Error {
-        if let Some(self_val) = self.err.downcast_ref::<Error>() {
+        let any_ref = (*self.err).as_any();
+
+        type_to_trait!(&dyn AnyError, Printable);
```
This is incorrect. `type_to_trait!` must be used only once for a type. i.e., as far as this PR is concerned, just in the test, for `PrintableErr`.

omkar-mohanty commented on src/result.rs:
```diff
@@ -47,6 +49,20 @@ impl Display for Error {
+
+impl<T: std::error::Error + Send + Sync + 'static> AnyError for T {}
+
+crate::impl_printable_for_display!(dyn AnyError);
```
You're right, that impl wasn’t needed — I’ve removed it now. Thanks for pointing it out!

vaivaswatha commented:
The CI is failing.

omkar-mohanty commented:
> The CI is failing.

It was a conflicting documentation on `std::error::Error` and `thiserror::Error`. It should be fixed now

vaivaswatha commented:
Thank you @omkar-mohanty !. looks good now. I'll merge it soon.
